GitCommands/Git/Executable.cs

```diff
@@ -134,8 +147,25 @@ private void OnProcessExit(object sender, EventArgs eventArgs)
                     {
                         try
                         {
-                            var exitCode = _process.ExitCode;
+                            int exitCode = _process.ExitCode;
                             _logOperation.LogProcessEnd(exitCode);
+
+                            if (_throwOnErrorOutput)
+                            {
+                                string errorOutput = _process.StandardError.ReadToEnd().Trim();
+                                if (exitCode != 0 || errorOutput.Length > 0)
+                                {
+                                    ExternalOperationException ex
+                                        = new(command: _process.StartInfo.FileName,
+                                              _process.StartInfo.Arguments,
+                                              _process.StartInfo.WorkingDirectory,
+                                              exitCode,
+                                              new Exception(errorOutput));
+                                    _logOperation.LogProcessEnd(ex);
+                                    _exitTaskCompletionSource.TrySetException(ex);
+                                }
+                            }
```

Comment:
Do we have tests for this?

Reply:
This is an integration. What would be the focal point of this test?

Reply:
Now looking at it, I don't think it is correct.

I want to fail on "exit code != 0", but I don't want to fail on "errorOutput.Length > 0" because it contains warnings.
If anything it has to be 
```cs
if (exitCode != 0 || (_throwOnErrorOutput && errorOutput.Length > 0))
```

As far as a test goes, we have integration tests in `ExecutableExtensionsTests` that allow testing this functionality.

Reply:
The intention with this PR is that errors and warnings should be handled, so checking the error output is in line with this.
In #9468 the throw will suppress both exit code and stderr, only show popup for exit code - could be popup on filtered stderr too.)
The origin of #9468 was that Git added printout to stderr in situations where the user did not expect them or see them before, so this should maybe be removed before a public release.
I suggest open an issue in milestone 3.6 for this and we catch warnings until then.

Reply:
Throwing on only stderr could be a separate argument too.

Reply:
By fixing one thing we must not break existing working things.

And I maintain this implementation is not correct - "exit code=0" implies the operation completed successfully. `errorOutput` may contain warnings, but they are not fatal, otherwise they'd been made errors.

Reply:
> By fixing one thing we must not break existing working things.

This PR **will** break functionality that happened to work before. See my conversation in this PR and PRs pointing to this as examples. (Of course, mstv fixed most problems in the PR but there is no history of that other than the code changes).

I find this breaking change important to find issues (at least during development), some to be suppressed, some to be suppressed, others to be fixed.
For production the popup could be suppressed (I currently want to have it enabled though.)

> And I maintain this implementation is not correct - "exit code=0" implies the operation completed successfully. errorOutput may contain warnings, but they are not fatal, otherwise they'd been made errors.

The warnings on stderr could be fatal, it is case by case. This could be yet another configuration.
But letting the caller make the choice is likely sufficient, similar to all other commands.

`throwOnErrorOutput` should not be allowed with `GetOutput*`, stderr should never be parsed as stdout.
Also, `includeErrorOutput` should be removed as these commands should fail.
This affects like 7 methods

`GetOutputLinesAsync` should be removed too (9 usages), will not be usable in #9476 anyway.
Not sure how the async yielding should be handled though.

Reply:
Draft cleanup in 3d320d33b8987b522b12dc88ea492ebc7ed84e80
The only actual change is that stderr is not parsed as an output (except SystemEncodingReader where I was unsure).

`GetOutput` is the simple flow where neither error exit nor stderr is allowed. 
"Advanced" usage with `Execute`

Reply:
> this implementation is not correct

No, it is. But the definition seems to be too strict.

> "exit code=0" implies the operation completed successfully. errorOutput may contain warnings, but they are not fatal, otherwise they'd been made errors.

I agree, this is a good definition.

Anyway, I want to get noticed of git warnings and non-fatal errors. They do not need to popup interrupting the command execution. The `ErrorOutput` should be added to the Git Command Log. New warnings should be indicated e.g. in the toolbar of FormBrowse or might be shown as Windows notifications. (Might need a configurable filter.)

PS: Regarding my status: It means tomorrow.

Reply:
`GetOutput()` as a simple method could well fail on any error output. stderr should normally not be parsed in the expected output.
I reviewed the use of stderr/AllOuput in #9488.
If `GetOutput` is changed to not throw if error exit is set, the method should only return stderr, not merge it with stderr.
`GetOutput` could be replaced globally too.